File: typescript/dnd-character/dnd-character.ts
export class DnDCharacter {
  readonly strength: number
  readonly dexterity: number
  readonly constitution: number
  readonly intelligence: number
  readonly wisdom: number
  readonly charisma: number
  readonly hitpoints: number

  constructor() {
    this.strength = DnDCharacter.generateAbilityScore()
    this.dexterity = DnDCharacter.generateAbilityScore()
    this.constitution = DnDCharacter.generateAbilityScore()
    this.intelligence = DnDCharacter.generateAbilityScore()
    this.wisdom = DnDCharacter.generateAbilityScore()
    this.charisma = DnDCharacter.generateAbilityScore()
    this.hitpoints = 10 + DnDCharacter.getModifierFor(this.constitution)
  }
  
  public static generateAbilityScore(): number {
    let roll = [];
    let sum = 0;
    for (let i = 0; i < 4;i++){
      let rand = Math.floor(Math.random() * 6 + 1)
      roll.push(rand);
      sum += rand
    }
    return sum - Math.min(...roll)
  }

  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2);
  }
}

File: solutions/go/dnd-character/1/dnd_character.go
package dndcharacter

import (
	"math"
	"math/rand"
)

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return 0
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	return 0
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	c := Character{
		Strength:     rand.Intn(15) + 3,
		Dexterity:    rand.Intn(15) + 3,
		Constitution: rand.Intn(15) + 3,
		Intelligence: rand.Intn(15) + 3,
		Wisdom:       rand.Intn(15) + 3,
		Charisma:     rand.Intn(15) + 3,
		Hitpoints:    0,
	}
	c.Hitpoints = int(math.Round(float64(c.Constitution - 10) / 2))
	return c
}
